alacritty_terminal/src/ansi.rs

```diff
@@ -25,60 +25,88 @@ use vte;
 use crate::term::color::Rgb;
 
 // Parse color arguments
-//
-// Expect that color argument looks like "rgb:xx/xx/xx" or "#xxxxxx"
-fn parse_rgb_color(color: &[u8]) -> Option<Rgb> {
-    let mut iter = color.iter();
 
-    macro_rules! next {
-        () => {
-            iter.next().map(|v| *v as char)
-        };
+// Parse a color argument in the form "#xxxxxx"
+// The use of this format is no longer encouraged by xparsecolor
+fn parse_legacy_rgb_color(color: &[u8]) -> Option<Rgb> {
+    if color.len() < 7 || color[0] != b'#' {
+        return None;
     }
 
+    let mut iter = color.iter().map(|v| (*v as char));
+    iter.next();
+
     macro_rules! parse_hex {
         () => {{
-            let mut digit: u8 = 0;
-            let next = next!().and_then(|v| v.to_digit(16));
-            if let Some(value) = next {
-                digit = value as u8;
-            }
+            let mut hex: u8 = match iter.next().and_then(|v| v.to_digit(16)) {
+                Some(digit) => digit as u8,
+                _ => return None,
+            };
+            hex <<= 4;
+            hex += match iter.next().and_then(|v| v.to_digit(16)) {
+                Some(digit) => digit as u8,
+                _ => return None,
+            };
+            hex
+        }};
+    }
+
+    let r = parse_hex!();
+    let g = parse_hex!();
+    let b = parse_hex!();
 
-            let next = next!().and_then(|v| v.to_digit(16));
-            if let Some(value) = next {
-                digit <<= 4;
-                digit += value as u8;
+    Some(Rgb { r, g, b })
+}
+
+fn is_hex_separator(ch: Option<char>) -> bool {
+    ch == None || ch == Some('/') || ch == Some('\x07')
+}
+
+// Expect that color argument looks like "rgb:x[xxx]/x[xxx]/x[xxx]" or "#xxxxxx"
+fn parse_rgb_color(color: &[u8]) -> Option<Rgb> {
+    let mut iter = color.iter().map(|v| *v as char);
+
+    macro_rules! parse_hex {
+        () => {{
+            let mut hex: u8 = match iter.next().and_then(|v| v.to_digit(16)) {
+                Some(digit) => digit as u8,
+                _ => return None,
+            };
+            hex <<= 4;
+
+            let next = iter.next();
+            if !is_hex_separator(next) {
+                if let Some(digit) = next.and_then(|v| v.to_digit(16)) {
+                    hex += digit as u8;
+                }
+                if (0..3).find(|_| is_hex_separator(iter.next())) == None {
+                    return None;
+                }
             }
-            digit
+
+            hex
         }};
     }
 
-    match next!() {
+    match iter.next() {
         Some('r') => {
-            if next!() != Some('g') {
+            if iter.next() != Some('g') {
                 return None;
             }
-            if next!() != Some('b') {
+            if iter.next() != Some('b') {
                 return None;
             }
-            if next!() != Some(':') {
+            if iter.next() != Some(':') {
                 return None;
             }
```

Comment:
I think this would be a worthwhile improvement:

```rust
let chars: Vec<u8> = iter.take(3).map(|c| *c as u8).collect();
if std::str::from_utf8(&chars) != Ok("gb:") {
    return None;
}
```

Reply:
I went for `len < 9 || &color[..4] != b"rgb:"` instead, which is similar to the legacy color approach - check the length and then check the beginning of the string before initializing the iterator.